### PR TITLE
Ra 489

### PR DESCRIPTION
**Pull request by vguzmanp**
This PR includes changes by aoller, with my fixes squashed.

It also includes the LimitationWidget story, it is already reviewed, but it would be nice to review it with the whole context.

### CODE REVIEW

File: VAS.Core/Events/StorageEvents.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+//  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
+//
+
+namespace VAS.Core.Events
+{
+	public abstract class StorageBaseEvent<T> : GenericEvent<T>
+	{
+	}
+
+	public class StorageAddEvent<T> : StorageBaseEvent<T>
```
Comment (ylatuya) [inline VAS.Core/Events/StorageEvents.cs]: Use the correct tense Added vs Add, it's an event to nitify about an accion, not to perform one

File: VAS.Core/Interfaces/ILicenseLimitationsService.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
+//
+using System;
+using VAS.Core.License;
+
+namespace VAS.Core.Interfaces
+{
+	public interface ILicenseLimitationsService : IService
```
Comment (ylatuya) [inline VAS.Core/Interfaces/ILicenseLimitationsService.cs]: Missing documentation

File: VAS.Core/Events/StorageEvents.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
+//
+
+namespace VAS.Core.Events
+{
+	public abstract class StorageBaseEvent<T> : GenericEvent<T>
```
Comment (ylatuya) [inline VAS.Core/Events/StorageEvents.cs]: Missing docs

File: VAS.Core/Interfaces/MVVMC/IViewModel.cs [hunk 4]
```diff
@@ -49,6 +49,12 @@ public interface INestedViewModel
 	/// </summary>
 	public interface INestedViewModel<VMChilds> : INestedViewModel, IViewModel, IEnumerable<VMChilds>
 	{
+		/// <summary>
+		/// Gets the collection that other classes should bind to
+		/// </summary>
+		/// <value>This should always be a calculated property referring to the real collection.</value>
+		RangeObservableCollection<VMChilds> BindableViewModelsCollection { get; }
```
Comment (ylatuya) [inline VAS.Core/Interfaces/MVVMC/IViewModel.cs]: I don't understand the need for this

File: VAS.Core/License/ILicenseLimitations.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+//
+using System;
+using System.Collections.Generic;
+
+namespace VAS.Core.License
+{
+	/// <summary>
+	/// License limitations Interface.
+	/// </summary>
+	public interface ILicenseLimitations
```
Comment (ylatuya) [inline VAS.Core/License/ILicenseLimitations.cs]: Interfaces must go the intefaces folder

File: VAS.Core/License/LicenseLimitation.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+//  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
+//
+using System;
+using VAS.Core.MVVMC;
+
+namespace VAS.Core.License
+{
+	/// <summary>
+	/// License limitation.
+	/// THis class represent a generic maxim limitation with enabled flag.
```
Comment (ylatuya) [inline VAS.Core/License/LicenseLimitation.cs]: typos

File: VAS.Core/License/ILicenseLimitations.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+		/// Adds the limitation.
+		/// </summary>
+		/// <param name="limitation">Limitation.</param>
+		void AddLimitation (LicenseLimitation limitation);
+
+		/// <summary>
+		/// Gets the limitations.
+		/// </summary>
+		/// <returns>The limitations.</returns>
+		IEnumerable<LicenseLimitation> GetLimitations ();
```
Comment (ylatuya) [inline VAS.Core/License/ILicenseLimitations.cs]: ILicenseLimitations should inherit from IEnumerable instead of adding this extra method.

File: VAS.Core/License/ILicenseLimitations.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+		/// </summary>
+		/// <returns>The limitations.</returns>
+		/// <param name="limitationName">Limitation name.</param>
+		LicenseLimitation GetLimitation (string limitationName);
+
+		/// <summary>
+		/// Sets the limitations status.
+		/// </summary>
+		/// <param name="status">If set to <c>true</c> status.</param>
+		void SetLimitationsStatus (bool status);
```
Comment (ylatuya) [inline VAS.Core/License/ILicenseLimitations.cs]: This should be a property Enabled

File: VAS.Core/License/ILicenseLimitations.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	/// <summary>
+	/// License limitations Interface.
+	/// </summary>
+	public interface ILicenseLimitations
+	{
+		/// <summary>
+		/// Adds the limitation.
+		/// </summary>
+		/// <param name="limitation">Limitation.</param>
+		void AddLimitation (LicenseLimitation limitation);
```
Comment (ylatuya) [inline VAS.Core/License/ILicenseLimitations.cs]: AddLimitation is redundat, just Add. Same for the rest of methods.

File: VAS.Core/License/ILicenseLimitations.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+		/// </summary>
+		/// <param name="status">If set to <c>true</c> status.</param>
+		void SetLimitationsStatus (bool status);
+
+		/// <summary>
+		/// Sets the limitation status.
+		/// </summary>
+		/// <param name="limitationName">Limitation name.</param>
+		/// <param name="status">If set to <c>true</c> status.</param>
+		void SetLimitationStatus (string limitationName, bool status);
```
Comment (ylatuya) [inline VAS.Core/License/ILicenseLimitations.cs]: rename "status" to "enabled", it makes more clear that we are enabling/disabling a limitiation. Not sure if this is really needed in the interface, because we can always iterate and filter by name.

File: VAS.Core/License/LicenseLimitation.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	/// </summary>
+	public class LicenseLimitation : BindableBase
+	{
+		/// <summary>
+		/// Gets or sets the name of the limitation.
+		/// </summary>
+		/// <value>The name of the limitation.</value>
+		public string LimitationName { get; set; }
```
Comment (ylatuya) [inline VAS.Core/License/LicenseLimitation.cs]: Just Name

File: VAS.Core/License/ILicenseLimitations.cs (hunk 5 above)
Comment (ylatuya) [inline VAS.Core/License/ILicenseLimitations.cs]: I don't really understand this interface, why can't we just use a IDiccionary<string, LicenseLimitation>

File: VAS.Core/MVVMC/LimitedCollectionViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+			if (Limitation?.Enabled ?? false) {
+				viewmodels = viewmodels.Take (Limitation.Maximum);
+			}
+
+			LimitedViewModels.Replace (viewmodels);
+		}
+
+		void LimitationPropertyChanged (object sender, PropertyChangedEventArgs e)
+		{
+			if (e.PropertyName == "Enabled" || e.PropertyName == "Maximum") {
```
Comment (ylatuya) [inline VAS.Core/MVVMC/LimitedCollectionViewModel.cs]: use nameof

File: VAS.Core/MVVMC/LimitedCollectionViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+				FillLimitedViewModels (ViewModels);
+			} else if (e.PropertyName == nameof (Limitation.Count)) {
+				Console.WriteLine ("Count changed");
```
Comment (ylatuya) [inline VAS.Core/MVVMC/LimitedCollectionViewModel.cs]: WriteLine

File: VAS.Core/ViewModel/LicenseLimitationVM.cs [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+	/// </summary>
+	public class LicenseLimitationVM : ViewModelBase<LicenseLimitation>
+	{
+		Command upgradeCommand;
+
+		public LicenseLimitationVM ()
+		{
+			UpgradeCommand = new Command (HandleUpgradeClicked) {
+				Icon = Resources.LoadIcon ("vas-limit-upgrade", 24),
+				ToolTipText = Catalog.GetString ("Freemium users can only create a limited number of projects at a time. " +
```
Comment (ylatuya) [inline VAS.Core/ViewModel/LicenseLimitationVM.cs]: This text can't be in VAS, each application will have to use a different one

File: VAS.Core/ViewModel/LicenseLimitationVM.cs [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+			get {
+				return base.Model;
+			}
+			set {
+				if (base.Model != null) {
+					base.Model.PropertyChanged -= HandleModelChanged;
+				}
+				base.Model = value;
+				if (base.Model != null) {
+					base.Model.PropertyChanged += HandleModelChanged;
```
Comment (ylatuya) [inline VAS.Core/ViewModel/LicenseLimitationVM.cs]: There is no need to connect signals, just override forwardpropertychanged

File: VAS.Core/ViewModel/LicenseLimitationVM.cs [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+		void HandleModelChanged (object sender, PropertyChangedEventArgs e)
+		{
+			if (e.PropertyName == nameof (Model.Enabled)) {
+				UpgradeCommand.Executable = Enabled;
+			}
+		}
+
+		void HandleUpgradeClicked ()
+		{
+			System.Diagnostics.Process.Start ("http://www." + App.Current.SoftwareName + ".com");
```
Comment (ylatuya) [inline VAS.Core/ViewModel/LicenseLimitationVM.cs]: This can't be hardcoded.

File: VAS.Core/ViewModel/TemplateViewModel.cs [hunk 17]
```diff
@@ -156,6 +157,12 @@ public INotifyCollectionChanged GetSelectionNotifyCollection ()
 			}
 		}
 
+		public RangeObservableCollection<TChildViewModel> BindableViewModelsCollection {
```
Comment (ylatuya) [inline VAS.Core/ViewModel/TemplateViewModel.cs]: Why is this needed?

File: VAS.DB/Views/GenericView.cs [hunk 18]
```diff
@@ -228,6 +228,33 @@ public IEnumerable<TBase> QueryFull (QueryFilter filter, IStorableObjectsCache c
...
+		/// <summary>
+		/// Performs a Count on the view with a <see cref="QueryFilter"/> whose keys
+		/// must be in the list of <see cref="FilterProperties"/> returning the count.
+		/// </summary>
+		/// <param name="filter">Filter.</param>
+		public int Count (QueryFilter filter)
+		{
+			View view = GetView ();
+
+			view.SetMapReduce (GetMap (DocumentType), (keys, values, rereduce) => {
```
Comment (ylatuya) [inline VAS.DB/Views/GenericView.cs]: This must be done inside GetView, no every single time we count, the reduce operation is only permformed on demand

File: VAS.Services/LicenseLimitationsService.cs [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+		public void SetCountValueForLimitation (string limitationName, int value)
+		{
+			var limit = LicenseLimitations.GetLimitation (limitationName);
+			if (limit != null) {
+				limit.Count = value;
+			}
+		}
+
+		public virtual void UpdateLicenseLimitationsCounters ()
```
Comment (ylatuya) [inline VAS.Services/LicenseLimitationsService.cs]: Why is this needed in the service interface and public ?

File: VAS.Services/LicenseLimitationsService.cs [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+		{
+			return true;
+		}
+
+		public virtual bool Stop ()
+		{
+			return true;
+		}
+
+		public void SetCountValueForLimitation (string limitationName, int value)
```
Comment (ylatuya) [inline VAS.Services/LicenseLimitationsService.cs]: Why is this even needed in the service and public? The one creating and setting the LicenseLimitations must create them with the correct counters

File: VAS.Core/License/LicenseLimitation.cs (hunk 6 above)
Comment (aameijide) [inline VAS.Core/License/LicenseLimitation.cs]: typo `THis`

File: VAS.Core/License/LicenseLimitations.cs [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="LicenseLimitation:VAS.Core.License.LicenseLimitations`1"/> class.
+		/// </summary>
+		public LicenseLimitations ()
+		{
+			Limitations = new Dictionary<string, LicenseLimitation> ();
+		}
+
+		public void AddLimitation (LicenseLimitation limitation)
```
Comment (aameijide) [inline VAS.Core/License/LicenseLimitations.cs]: Document public methods

File: VAS.DB/CouchbaseStorage.cs [hunk 22]
```diff
@@ -183,13 +183,22 @@ public void Fill (IStorable storable)
 			}
 		}
 
+		public int Count<T> (QueryFilter filter) where T : IStorable
```
Comment (aameijide) [inline VAS.DB/CouchbaseStorage.cs]: Document method

File: VAS.DB/Views/GenericView.cs [hunk 23]
```diff
@@ -228,6 +228,33 @@ public IEnumerable<TBase> QueryFull (QueryFilter filter, IStorableObjectsCache c
...
+			}, ViewVersion);
+
+			Query q = view.CreateQuery ();
+			q.MapOnly = false;
+
+			q.SQLSearch = QueryFilterToSql (filter);
+			QueryEnumerator ret = q.Run ();
+			int count = 0;
+			if (ret.Count > 0) {
+				count = Convert.ToInt32 (ret.ElementAt (0).Value);
```
Comment (aameijide) [inline VAS.DB/Views/GenericView.cs]: This can be changed to linq equivalent
`if(ret.Any()) { count = Convert.ToInt32(ret.FirstOrDefault().Value); }`that in my opinion is clearer to read

File: VAS.Services/LicenseLimitationsService.cs [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+			var limit = LicenseLimitations.GetLimitation (limitationName);
+			if (limit != null) {
+				limit.Count = value;
+			}
+		}
+
+		public virtual void UpdateLicenseLimitationsCounters ()
+		{
+		}
+	}
```
Comment (aameijide) [inline VAS.Services/LicenseLimitationsService.cs]: Document public properties

File: VAS.Tests/Core/License/TestLicenseLimitations.cs [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+		[Test ()]
+		public void TestAddLimitations ()
+		{
+			limitations.AddLimitation (limitationPlayers);
+			limitations.AddLimitation (limitationPlayers2);
+			limitations.AddLimitation (limitationTeams);
+
+			LicenseLimitation testLimitationPlayers = limitations.GetLimitation ("RAPlayers");
+			LicenseLimitation testLimitationTeams = limitations.GetLimitation ("Teams");
+			List<LicenseLimitation> allLlimitations =
```
Comment (aameijide) [inline VAS.Tests/Core/License/TestLicenseLimitations.cs]: Split each action in a new test

**Comment by ylatuya** [comment]
Overall, I think are several security concerns that make this implementation very fragile. We should ensure the following:
  * The Limitations service should be accesible to views or plugins. 
  * Limitations can't be removed at any point by anyone.
  * Limitations status shouldn't be modifiable.
  * Limitations limits shouldn't be modifiable.

We'll need to discuss all these aspects in detail.

File: VAS.Core/Interfaces/MVVMC/IViewModel.cs (hunk 4 above)
Comment (vguzmanp) [inline VAS.Core/Interfaces/MVVMC/IViewModel.cs]: This is only actually used in the `LimitedCollectionViewModel`. It was needed to "override" the ViewModels collection with a limited collection. Using a real override quickly became a hell to manage.

File: VAS.Core/ViewModel/LicenseLimitationVM.cs (hunk 16 above)
Comment (vguzmanp) [inline VAS.Core/ViewModel/LicenseLimitationVM.cs]: You mean the url or the whole upgrade behaviour?

**Comment by vguzmanp** [comment]
@ylatuya Answering to the security comment:

- The Limitations service is accesible through App.Current.LimitationService, and it's a Service registered to the global Service register. Do we need anything else to make it accesible to views / plugins?

- We talked about making the app react to changes in the limitations. When it's implemented, this will be read from the protection system, but at the moment we can't read them, so we need to write it somewhere, right?

File: VAS.Core/Events/StorageEvents.cs (hunk 3 above)
Comment (vguzmanp) [inline VAS.Core/Events/StorageEvents.cs]: ☑️

File: VAS.Core/Events/StorageEvents.cs (hunk 1 above)
Comment (vguzmanp) [inline VAS.Core/Events/StorageEvents.cs]: ☑️

File: VAS.DB/Views/GenericView.cs (hunk 18 above)
Comment (vguzmanp) [inline VAS.DB/Views/GenericView.cs]: Fixed in d109b6cce1b244e52cfb411f970eb64d004e1542. I'm not sure if you meant that.

Comment (vguzmanp) [inline VAS.DB/Views/GenericView.cs]: Actually, moving this to `GetView` breaks everything else. All DB-related tests fail. Is that change what you meant?

File: VAS.Tests/Services/TestLicenseLimitationService.cs [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+		[Test]
+		public void TestGetNonExisting ()
+		{
+			LicenseLimitationVM limit = service.Get ("Non-existing limitation");
+			IEnumerable<LicenseLimitationVM> allLimitations = (service as DummyLicenseLimitationsService).GetAll ();
+
+			Assert.AreEqual (0, allLimitations.Count ());
+			Assert.IsNull (limit);
+		}
+	}
```
Comment (aameijide) [inline VAS.Tests/Services/TestLicenseLimitationService.cs]: Add a test getting limitation count when is disabled

File: VAS.Core/ViewModel/TemplateViewModel.cs [hunk 27]
```diff
@@ -156,6 +157,12 @@ public INotifyCollectionChanged GetSelectionNotifyCollection ()
...
+			get {
+				return ViewModels;
```
Comment (aameijide) [inline VAS.Core/ViewModel/TemplateViewModel.cs]: Shouldn't this return a limited VM?

File: VAS.Core/License/LicenseLimitation.cs [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+		/// <value>The count.</value>
+		public int Count { get; set; }
+
+		/// <summary>
+		/// Gets or sets the maximum number for the limitation.
+		/// </summary>
+		/// <value>The maximum.</value>
+		public int Maximum {
+			get {
+				// FIXME: Logic in the model?
```
Comment (vguzmanp) [inline VAS.Core/License/LicenseLimitation.cs]: Yes, this is a doubt about what to do. I don't like having logic in the model, but using a controller for this is way too much of overhead.

File: VAS.Core/MVVMC/NestedViewModel.cs [hunk 29]
```diff
@@ -39,6 +39,18 @@ public NestedViewModel ()
 			ViewModels.CollectionChanged += HandleViewModelsChanged;
 		}
 
+
+		/// <summary>
+		/// Gets the collection that other classes should bind to.
+		/// This should always be a calculated property referring to the real collection.
+		/// </summary>
+		//FIXME: This name is shit, any other good one?
```
Comment (vguzmanp) [inline VAS.Core/MVVMC/NestedViewModel.cs]: I don't follow you here
